openml/flows/sklearn_converter.py

```diff
@@ -0,0 +1,466 @@
+
+MAXIMAL_FLOW_LENGTH = 1024
+
+
+if sys.version_info >= (3, 5):
+    from json.decoder import JSONDecodeError
+else:
+    JSONDecodeError = ValueError
+
+
+class SklearnToFlowConverter(object):
+    """Convert scikit-learn estimator into an OpenMLFlow."""
+
+    def serialize_object(self, o):
+
+        if self._is_estimator(o) or self._is_transformer(o):
+            rval = self._serialize_model(o)
+        elif isinstance(o, (list, tuple)):
+            rval = [self.serialize_object(element) for element in o]
+            if isinstance(o, tuple):
+                rval = tuple(rval)
+        elif o is None:
+            rval = None
+        elif isinstance(o, six.string_types):
+            rval = o
+        elif isinstance(o, bool):
+            # The check for bool has to be before the check for int, otherwise,
+            # isinstance will think the bool is an int and convert the bool will
+            # be converted to a string which can't be parsed by json.loads
+            rval = o
+        elif isinstance(o, int):
+            rval = o
+        elif isinstance(o, float):
+            rval = o
+        elif isinstance(o, dict):
+            rval = {}
+            for key, value in o.items():
+                if not isinstance(key, six.string_types):
+                    raise TypeError('Can only use string as keys, you passed '
+                                    'type %s for value %s.' %
+                                    (type(key), str(key)))
+                key = self.serialize_object(key)
+                value = self.serialize_object(value)
+                rval[key] = value
+            rval = rval
+        elif isinstance(o, type):
+            rval = self.serialize_type(o)
+        elif isinstance(o, scipy.stats.distributions.rv_frozen):
+            rval = self.serialize_rv_frozen(o)
+        # This only works for user-defined functions (and not even partial).
+        # I think this exactly we want here as there shouldn't be any built-in
+        # or functool.partials in a pipeline
+        elif inspect.isfunction(o):
+            rval = self.serialize_function(o)
+        elif self._is_cross_validator(o):
+            rval = self.serialize_cross_validator(o)
+        else:
+            raise TypeError(o, type(o))
+
+        assert o is None or rval is not None
+
+        return rval
+
+    # TODO maybe remove those functions and put the check to the long
+    # if-constructs above?
+    def _is_estimator(self, o):
+        # TODO @amueller which checks is preferable?
+        # return (hasattr(o, 'fit') and hasattr(o, 'predict') and
+        #         hasattr(o, 'get_params') and hasattr(o, 'set_params'))
+        return isinstance(o, sklearn.base.BaseEstimator)
+
+    def _is_transformer(self, o):
+        # TODO @amueller should one rather check whether this is a subclass of
+        # BaseTransformer?
+        return (hasattr(o, 'fit') and hasattr(o, 'transform') and
```

Comment:
If it's not an estimator, it will go into the else branch where it raises a `TypeError`. Do you mean I should add a more descriptive error message there?